api/envoy/extensions/filters/network/mysql_proxy/v3/mysql_proxy.proto

```diff
@@ -15,15 +18,71 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // MySQL Proxy :ref:`configuration overview <config_network_filters_mysql_proxy>`.
 // [#extension: envoy.filters.network.mysql_proxy]
 
+// Auth information of MySQL.
+message AuthInfo {
+  config.core.v3.DataSource auth_username = 1;
+
+  config.core.v3.DataSource auth_password = 2;
+}
+
+// [#next-free-field: 6]
 message MySQLProxy {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.network.mysql_proxy.v1alpha1.MySQLProxy";
 
+  message Route {
+    // String database name that must exact match with client connection db name
+    string database = 1 [(validate.rules).string = {min_len: 1}];
+
+    // Upstream cluster to forward the command to.
+    string cluster = 2 [(validate.rules).string = {min_len: 1}];
+  }
+
   // The human readable prefix to use when emitting :ref:`statistics
   // <config_network_filters_mysql_proxy_stats>`.
   string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
 
   // [#not-implemented-hide:] The optional path to use for writing MySQL access logs.
   // If the access log field is empty, access logs will not be written.
   string access_log = 2;
+
+  // [#not-implemented-hide:] enable feature of separate authentication of downstream clients and upstream server. When feature_enable is true, the filter is a terminal filter, it's config should put at the last of filter chain, when feature_enable is false, the filter is not a terminal filter, it's config can not put at the last of filter chain. Default is false.
+  bool feature_enable = 3;
+
+  // List of **unique** db name used to separate connections from different database to different
+  // clusters.
+  //
+  // Example:
+  //
+  // .. code-block:: yaml
+  //
+  //    routes:
+  //      - database: "database_a"
+  //        cluster: "cluster_a"
+  //      - database: "database_b"
+  //        cluster: "cluster_b"
+  //
+  // When using the above routes, the following connect dsn would be sent to:
+  //
+  // * ``jdbc:mysql://host1:9001/database_a`` would sent command to cluster_a.
+  // * ``jdbc:mysql://host1:9001/database_c`` will return ER_ER_BAD_DB_ERROR.
+  // * ``jdbc:mysql://host1:9001/`` will return ER_ER_BAD_DB_ERROR, now envoy not support connect without specific database name.
+  repeated Route routes = 4;
+
+  // Downstream client username/password.
+  // * If username and password is empty, then authentication will always pass.
+  // * If username is empty but password is not, then proxy will only auth password.
+  // * If password is empty but username is not, then proxy will only auth username/
+  // * If the username is not match, then ER_ACCESS_DENIED_ERROR will return;
+  // * If connect db name is not found in envoy.extensions.filters.network.mysql_proxy.v3.MySQLProxy.routes, then ER_ER_BAD_DB_ERROR will return.
```

Comment:
What does the last bullet point have to do with auth?

Reply:
removed.